src/main/java/dev/morling/onebrc/CalculateAverage_plbpietrz.java

```diff
@@ -0,0 +1,211 @@
+        Map<Integer, TemperatureStats> results;// = new HashMap<>();
+        try (RandomAccessFile inputFile = new RandomAccessFile(inputFilePath.toFile(), "r")) {
+            var parsedBuffers = partitionInput(inputFile)
+                    .stream()
+                    .parallel()
+                    .map(fp -> getMappedByteBuffer(fp, inputFile))
+                    .map(CalculateAverage_plbpietrz::parseBuffer);
+            results = parsedBuffers.flatMap(m -> m.entrySet().stream())
+                    .collect(
+                            Collectors.groupingBy(
+                                    Map.Entry::getKey,
+                                    Collectors.reducing(
+                                            new TemperatureStats(),
+                                            Map.Entry::getValue,
+                                            CalculateAverage_plbpietrz::mergeTemperatureStats)));
+            try (PrintWriter pw = new PrintWriter(new BufferedOutputStream(System.out))) {
+                formatResults(pw, results);
+            }
+        }
+    }
+
+    private static List<FilePart> partitionInput(RandomAccessFile inputFile) throws IOException {
+        List<FilePart> fileParts = new ArrayList<>();
+        long fileLength = inputFile.length();
+
+        long blockSize = Math.min(fileLength, Math.max(1024, fileLength / CPU_COUNT));
+
+        for (long start = 0, end; start < fileLength; start = end) {
+            end = findMinBlockOffset(inputFile, start, blockSize);
+            fileParts.add(new FilePart(start, end - start));
+        }
+        return fileParts;
+    }
+
+    private static long findMinBlockOffset(RandomAccessFile file, long startPosition, long minBlockSize) throws IOException {
+        long length = file.length();
+        if (startPosition + minBlockSize < length) {
+            file.seek(startPosition + minBlockSize);
+            while (file.readByte() != '\n') {
+            }
+            return file.getFilePointer();
+        }
+        else {
+            return length;
+        }
+    }
+
+    private static MappedByteBuffer getMappedByteBuffer(FilePart fp, RandomAccessFile inputFile) {
+        try {
+            return inputFile.getChannel().map(FileChannel.MapMode.READ_ONLY, fp.pos, fp.size);
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private static Map<Integer, TemperatureStats> parseBuffer(MappedByteBuffer buffer) {
+        byte[] readLong = new byte[READ_SIZE];
+        byte[] stationName = new byte[512];
+        byte[] temperature = new byte[32];
+        int stationLineNameLenght = 0;
+        int temperatureLineLenght = 0;
+        int stationNameHash = 0;
+
+        int limit = buffer.limit();
+        boolean readingName = true;
+        Map<Integer, TemperatureStats> temperatures = new HashMap<>();
+        Map<Integer, String> stationNames = new HashMap<>();
```

Comment:
How do you deal with hash collisions here?

Reply:
Well, this code makes assumptions about the input data, specifically that the station names will not collide under the chosen hash function. If that happens, unfortunately corrupted output will be produced.

Reply:
Ok, that's not valid then though as per [the rules](https://github.com/gunnarmorling/1brc#faq):

> Q: Can I make assumptions on the names of the weather stations showing up in the data set?
A: No, while only a fixed set of station names is used by the data set generator, any solution should work with arbitrary UTF-8 station names (for the sake of simplicity, names are guaranteed to contain no ; or \n characters).